syntax_tools: Allow merl quote functions to accept an erl_anno:anno()

maxno-kivra opened the pull request:
Merl currently assumes all annotations are locations, which can be
either an integer or a two-tuple of integers. However, erl_anno
supports more fields, like file and text. This is done through an
opaque type, which is actually a proplist.

OTP adds support for column numbers, and it would be nice if
merl would support those somewhere down the line. A first step
would be to allow the quote function(s) to accept the full
erl_anno:anno() type. It currently pulls the location and then
continues as normal.

I also renamed Line to Location in merl_transform to make it clear
that it can be both a integer and tuple of integers. Later perhaps
a full anno, but not now..

---

I'm not sure what to call the test, but I noticed `otp_15291`, is that named after some issue number? Should I change the name to this PR's number?

garazdawi commented:
> I'm not sure what to call the test, but I noticed otp_15291, is that named after some issue number? Should I change the name to this PR's number?

You can leave it as is. Using the ticket numbers is an old convention that we have moved away from.

ranjanified commented:
Yes please. anno_test is much more meaningful than otp_1234****. :-) Names
are meant for us, to help each other. Otherwise it will be a journey to go
find out what is otp_123**** or will have to read through the test
implementation to infer anything.

On Thu, Jan 28, 2021 at 10:05 PM Lukas Larsson <notifications@github.com>
wrote:

> I'm not sure what to call the test, but I noticed otp_15291, is that named
> after some issue number? Should I change the name to this PR's number?
>
> You can leave it as is. Using the ticket numbers is an old convention that
> we have moved away from.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/erlang/otp/pull/3010#issuecomment-769210399>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABRD7V2VBHNUHZ6AKAGHZJ3S4GG6ZANCNFSM4WXLBFDA>
> .
>

maxno-kivra commented:
I had a bit of trouble getting dialyzer to run, and a bit of trouble convincing it that the code's fine. But switch to `erl_anno:from_term` fixes that. Let's see if the tests passes. Otherwise I might need some help.

maxno-kivra commented:
The tests, and dialyzer, passed. Hurray! I think this means it's ready.

uabboli commented:
I don't quite understand what you're trying to do.

merl:quote() is documented to accept a location() as first argument.
Why are you trying to change that? It doesn't make sense to me.

Note that erl_anno:location() is not an opaque type; you're allowed to
check if it's a tuple, an integer, &c.

Perhaps I'll understand better if you have an example. The test
anno_test() fails on master, but that is what it's supposed to do.

(If you choose continue this PR, you should rebase on master.)

michalmuskala commented:
> Note that erl_anno:location() is not an opaque type; you're allowed to
check if it's a tuple, an integer, &c.

I'm not sure this is accurate. The dialyzer type is defined with `-opaque` and the [documentation](https://erlang.org/doc/man/erl_anno.html#type-anno) doesn't specify the structure of the type.

maxno-kivra commented:
I've rebased on master now. The conflict was from #3006 and @richcarl work on using erl_anno opaquely. Here is the commit if you're curious https://github.com/erlang/otp/commit/f30514e3232f63563251a42334912aef8c79d4e8#diff-f2e48eb7504acfca64629944afb87c972e5430f81f047d1c5fff0df1b7651edaR1251-R1254.

Like that commit mentions, what used to be a line number is now a general storage of annotations. Like @michaelklishin pointed out, and the reason I first got dialyzer errors, is that `erl_anno:anno/0` is considered an opaque type, that usually happens to be a number. Sometimes it's a `{Line, Column}` tuple, and sometimes it's a keyword list.

Here's the definition of `new_location/1`. First off it's different with `DEBUG` set or not, but you can also see how locations are stored as a keyword list.

https://github.com/erlang/otp/blob/ad7b57a5dd1ef98f20b90e766e403ca06ebcd10a/lib/stdlib/src/erl_anno.erl#L118-L124

I'm trying to figure out how to compile with `DEBUG` set, to see what happens. @uabboli any pointers on how to do that?

maxno-kivra commented:
I've been thinking a bit about this one. This PR does not fix a bug, you're absolutely right @uabboli. What it does is pave the road towards supporting column numbers, and end location.

I've updated the commit message, title and description to reflect that. I've also rebased on latest master. What do you think?

maxno-kivra commented:
Since this hasn't been commented on I quite some time I think it's time to close it. If the need arises in the future it can always be reopened.
